virttest/ip_sniffing.py

```diff
@@ -0,0 +1,311 @@
+
+    def start(self):
+        """Start sniffing."""
+        if self.is_alive():
+            return
+        self.stop()
+        self._start()
+
+    def stop(self):
+        """Stop sniffing."""
+        if self._process:
+            self._process.close()
+            old_p = self._process
+            self._process = None
+            del old_p
+
+    def __del__(self):
+        self.stop()
+
+
+class TcpdumpSniffer(Sniffer):
+
+    """
+    Tcpdump sniffer class.
+    """
+
+    command = "tcpdump"
+    options = "-tnpvvvi any 'port 68 or port 546'"
+
+    def __init__(self, addr_cache, log_file, remote_opts=None):
+        super(TcpdumpSniffer, self).__init__(addr_cache, log_file, remote_opts)
+        self._context = {}
+
+    def _output_handler(self, line):
+        # BootP/DHCP (RFC 951/2131)
+        matches = re.search(r"^IP\s", line)
+        if matches:
+            # Clear context upon receiving a new packet
+            self._context.clear()
+            return
+
+        matches = re.search(r"Your.IP\s+(\S+)", line, re.I)
+        if matches:
+            self._context["ip"] = matches.group(1)
+            return
+
+        matches = re.search(r"Client.Ethernet.Address\s+(\S+)", line, re.I)
+        if matches:
+            self._context["mac"] = matches.group(1)
+            return
+
+        if re.search(r"DHCP.Message.*:\s+ACK", line, re.I):
+            mac = self._context.get("mac")
+            ip = self._context.get("ip")
+            if mac and ip:
+                self._cache[mac] = ip
+            return
+
+        # DHCPv6 (RFC 3315)
+        if re.search("dhcp6 reply", line, re.I):
+            regex = "IA_ADDR (.*) pltime.*client-ID.*?([0-9a-fA-F]{12})\)"
+            matches = re.search(regex, line, re.I)
+            if matches:
+                info = matches.groups()
+                mac = ":".join(re.findall("..", info[1]))
+                ip = info[0].lower()
+                self._cache["%s_6" % mac] = ip
+            return
+
+
+class TSharkSniffer(Sniffer):
+
+    """
+    TShark sniffer class.
+    """
+
+    command = "tshark"
+    options = ("-npi any 'port 68 or port 546' -T fields -E header=y "
+               "-E separator=/s -E occurrence=f -e ip.src -e ip.dst "
+               "-e bootp.type -e bootp.id -e bootp.hw.mac_addr "
+               "-e bootp.ip.your -e bootp.option.dhcp")
+
+    def _output_handler(self, line):
+        packet = line.split()
+        if not len(packet):
+            return
+
+        # BootP/DHCP (RFC 951/2131)
+        if re.match(r"\d+\.\d+\.\d+\.\d+", packet[0]):
+            chaddr = packet[4]
+            yiaddr = packet[5]
+            m_type = packet[6]
+            if m_type == "5" and yiaddr != "0.0.0.0":
+                # Update cache only if get the ACK reply
+                # and the previous request is not INFORM
+                self._cache[chaddr] = yiaddr
+            return
+
+        # DHCPv6 (RFC 3315)
+        # TODO: support DHCPv6
+        pass
```

Comment:
it doesn't need this pass

Reply:
I'd like to keep this line for pointing out the end of this function.

Reply:
How about adding regexp to detect DHCPv6 usage and log warning in such case? I mean if you are certain you'll finish the DHCPv6 support in a month, then sure. I'm only afraid this TODO will get lost and in 5 years we might be wondering why our sniffers don't work.

Anyway it's just a mild suggestion, so far it looks good, I'll test how it behaves so we can proceed.